README.md
# CSV-to-OpenImmo converter

[![Build Status](https://travis-ci.org/oliverklee/ext-csv_to_openimmo.svg?branch=main)](https://travis-ci.org/oliverklee/ext-csv_to_openimmo)
[![Latest Stable Version](https://poser.pugx.org/oliverklee/csv_to_openimmo/v/stable.svg)](https://packagist.org/packages/oliverklee/csv_to_openimmo)
[![Total Downloads](https://poser.pugx.org/oliverklee/csv_to_openimmo/downloads.svg)](https://packagist.org/packages/oliverklee/csv_to_openimmo)
[![Latest Unstable Version](https://poser.pugx.org/oliverklee/csv_to_openimmo/v/unstable.svg)](https://packagist.org/packages/oliverklee/csv_to_openimmo)
[![License](https://poser.pugx.org/oliverklee/csv_to_openimmo/license.svg)](https://packagist.org/packages/oliverklee/csv_to_openimmo)

This TYPO3 extension provides a Scheduler task that reads zipped CSV files
(created using
[Wodis Sigma](https://www.aareon.de/Produkte.5772.html?tx_aareon_selector[category]=5774))
from a configured folder and writes zipped OpenImmo files to another configured
folder. The task will also copy all image files and PDF from the ZIPs.

## Data format

You can find the CSV column definitions in the
[CSV Reader source code file](Classes/Service/CsvReader.php).

The functional tests also include an
[example CSV file](Tests/Functional/Fixtures/CorrectCsv/objects.csv).

The task will also automatically convert from ISO-9959-1 to UTF-8.

This converter will automatically mark all converted objects as "for rent".
Hence, it will not be able to convert objects that are for sale.

## Installing the extension

If you install the extension manually (not from the TER and not using Composer),
please make sure to name the extension directory `csv_to_openimmo`, not
`ext-csv_to_openimmo`.

## Caveats

The following OpenImmo fields currently are not imported:

- "vermarktungsart" (will always be "MIETE_PACHT")
- other image types than JPEG
- the type of transfer ("uebertragung > umfang", will always be "VOLL")

## Sponsorship

Development has been sponsored by
[ART-KON-TOR Kommunikation GmbH](https://www.art-kon-tor.de/).


CHANGELOG.md
# Change log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](https://semver.org/).

## x.y.z

### Added

### Changed
- Switch the default git branch from master to main (#110)
- Run the TER release script with PHP 7.2 (#106)
- Sort the Composer dependencies (#105)
- Simplify the Composer dependency versions (#101)
- Change from GPL V3+ to GPL V2+ (#95)

### Deprecated

### Removed
- Drop unneeded Travis CI configuration settings (#102, #103)
- Remove the TER as Composer repository (#100)
- Stop running with different dependency versions (#99)

### Fixed
- Stop packaging development-only files for Packagist (#111)
- Always use Composer-installed versions of the dev tools (#109)
- Fix warnings in the `.travis.yml` (#108)
- Do not cache `vendor/` on Travis CI (#107)
- Drop an obsolete "replace" entry from composer.json (#97)
- Properly quote the PHP version strings in .travis.yml (#96)
- Explicitly provide the extension name in the composer.json (#94)

## 2.0.0

### Added
- Add a test case with a minimal valid OpenImmo file (#88)

### Changed
- Streamline ext_emconf.php (#86)

### Removed
- Require TYPO3 >= 7.6 (#87)

### Fixed
- Also process ampersands in the CSV (#90, #91)
- Also provide the extension icon in `Resources/` (#89)

## 1.0.0

### Added
- Auto-release to the TER (#78)
- Convert more utilization types (#73, #74)
- Add compatibility with TYPO3 CMS 8.7 (#70)
- Convert the "habitation" utilization type (#68, #61)
- Always mark the object address as visible (#62, #63)
- Also convert "commercial use" (#31, #56)
- Convert the elevator field (#51, #43)
- Convert parking space types (#36, #42)
- Convert the number of balconies (#26, #40)
- Convert additional heating types (#33, #39)
- Convert the street number as an additional field (#32, #38)
- Add compatibility with PHP 5.5 (#23)

### Changed
- Prefer stable/dist packages by default (#80)
- Change the OpenImmo transfer type from "partial" to "full" (#65)
- Convert the floor as a free-text field (#48, #57)
- Also use "parking" for the object type (#46, #53)
- Convert balcony and patio as one field (#47, #52)
- Mark the import mode as "update" instead of "new" (#34, #41)

### Fixed
- Also create OpenImmo nodes for empty data (#84)
- Create valid OpenImmo for commercial objects (#75, #77)
- Convert only balconies, not balconies and patios combined (#76)
- Add the missing dependency on ext-libxml (#66)
- Add the missing dependency on ext-dom (#64)
- Always set all utilization types to false (#50, #49)


Classes/Service/CsvReader.php
<?php
namespace OliverKlee\CsvToOpenImmo\Service;

use TYPO3\CMS\Core\SingletonInterface;
use TYPO3\CMS\Core\Utility\GeneralUtility;

/**
 * This class reads and checks a CSV file with realty data.
 *
 * @author Oliver Klee <typo3-coding@oliverklee.de>
 */
class CsvReader implements SingletonInterface
{
    /**
     * @var bool
     */
    const DEBUG = false;

    /**
     * @var int
     */
    const NUMBER_OF_COLUMNS = 107;

    /**
     * The keys are one-based, not zero-based
     *
     * @var string[]
     */
    private static $columnKeys = [
        3 => 'utilization',
        4 => 'objectNumber',
        6 => 'street',
        7 => 'streetNumber',
        8 => 'zip',
        9 => 'city',
        10 => 'district',
        11 => 'yearOfConstruction',
        13 => 'numberOfFloors',
        14 => 'availabilityDate',
        15 => 'floor',
        16 => 'numberOfRooms',
        17 => 'livingArea',
        18 => 'heatingType',
        23 => 'rentWithoutHeatingCosts',
        24 => 'additionalCosts',
        26 => 'heatingIncludedInAdditionalCosts',
        27 => 'rentWithHeatingCosts',
        29 => 'deposit',
        30 => 'title',
        32 => 'location',
        33 => 'description',
        34 => 'equipment',
        35 => 'elevator',
        36 => 'balcony_or_patio',
        46 => 'parkingSpaceType',
        49 => 'contactPersonSalutation',
        50 => 'contactPersonFullName',
        55 => 'contactPersonPhoneNumber',
        56 => 'contactPersonEmail',
        58 => 'imageFileName 01',
        62 => 'imageTitle 01',
        63 => 'imageFileName 02',
        67 => 'imageTitle 02',
        68 => 'imageFileName 03',
        72 => 'imageTitle 03',
        73 => 'imageFileName 04',
        77 => 'imageTitle 04',
        78 => 'imageFileName 05',
        82 => 'imageTitle 05',
    ];

    /**
     * The keys are one-based, not zero-based
     *
     * @var string[]
     */
    private static $ignoredColumns = [
        1 => 'I',
        2 => '1',
        5 => 'company name',
        19 => '119,44',
        20 => 'Bedarfsausweis',
        21 => '08.20.2008',
        25 => '83',
        31 => 'full object address',
        37 => 'Kellerbox 401',
        51 => 'some street (of the offerer or contact person?)',
        52 => 'ZIP code (of the offerer or contact person?)',
        53 => 'city (of the offerer or contact person?)',
        54 => 'some phone number (whose? of the contact person?)',
        57 => 'some URL',
        59 => 'jpg',
        60 => 'B',
        64 => 'jpg',
        65 => 'B',
        69 => 'jpg',
        70 => 'B',
        74 => 'jpg',
        75 => 'B',
        79 => 'jpg',
        80 => 'B',
    ];

    /**
     * @param string $directory
     *
     * @return string|null the absolute path to the file
     */
    public function findFirstCsvFileInDirectory($directory)
    {
        /** @var string[] $allCsvFiles */
        $allCsvFiles = GeneralUtility::getAllFilesAndFoldersInPath([], $directory, 'csv');

        return array_shift($allCsvFiles);
    }

    /**
     * @param string $path
     *
     * @return string[][]
     *
     * @throws \RuntimeException
     * @throws \UnexpectedValueException
     */
    public function readCsv($path)
    {
        $rawCsvLines = $this->readCsvLines($path);
        return $this->convertLinesToSpeakingKeys($rawCsvLines);
    }

    /**
     * @param string $path
     *
     * @return string[][]
     *
     * @throws \RuntimeException
     * @throws \UnexpectedValueException
     */
    private function readCsvLines($path)
    {
        if (!is_file($path) || !is_readable($path)) {
            throw new \RuntimeException('The CSV file "' . $path . '" could not be read.', 1526493542);
        }

        /** @var string[][] $csvLines */
        $csvLines = [];
        $fileHandle = fopen($path, 'rb');
        while (($csvLine = fgetcsv($fileHandle, 0, ';'))) {
            /** @var string[] $csvLine */
            $csvLines[] = $csvLine;
        }
        fclose($fileHandle);

        $this->validateNumberOfColumns($path, $csvLines);

        return $csvLines;
    }

    /**
     * @param string $path
     * @param string[][] $csvLines
     *
     * @return void
     *
     * @throws \UnexpectedValueException
     */
    private function validateNumberOfColumns($path, array $csvLines)
    {
        if (!empty($csvLines)) {
            $numberOfColumns = count($csvLines[0]);
            if ($numberOfColumns !== self::NUMBER_OF_COLUMNS) {
                throw new \UnexpectedValueException(
                    'The CSV file "' . $path . '" should have ' . self::NUMBER_OF_COLUMNS .
                    ' columns, but actually has ' . $numberOfColumns . ' columns.',
                    1526494172
                );
            }
        }
    }

    /**
     * @param string[][] $rawCsvLines
     *
     * @return string[][]
     */
    private function convertLinesToSpeakingKeys(array $rawCsvLines)
    {
        $result = [];
        foreach ($rawCsvLines as $rawCsvLine) {
            $lineWithSpeakingKeys = $this->convertSingleLineToSpeakingKeys($rawCsvLine);
            $encodedLine = array_map('utf8_encode', $lineWithSpeakingKeys);
            $trimmedEncodedLine = array_map('trim', $encodedLine);
            $result[] = $trimmedEncodedLine;
        }

        return $result;
    }

    /**
     * @param string[] $rawLine
     *
     * @return string[]
     */
    private function convertSingleLineToSpeakingKeys(array $rawLine)
    {
        $result = [];

        foreach ($rawLine as $zeroBasedColumnNumber => $value) {
            /** @var int $zeroBasedColumnNumber */
            $oneBasedColumnNumber = $zeroBasedColumnNumber + 1;
            if (array_key_exists($oneBasedColumnNumber, self::$columnKeys)) {
                $speakingKey = self::$columnKeys[$oneBasedColumnNumber];
                $result[$speakingKey] = $value;
            } else {
                if (self::DEBUG && !array_key_exists($oneBasedColumnNumber, self::$ignoredColumns)
                    && trim($value) !== '') {
                    throw new \UnexpectedValueException(
                        'The column #' . $oneBasedColumnNumber . ' with the value "' .
                        utf8_encode($value) . '" is not mapped yet.',
                        1526554511
                    );
                }
            }
        }

        return $result;
    }
}
